nmoscommon/auth/auth_registrar.py

```diff
@@ -0,0 +1,200 @@
+# Copyright 2019 British Broadcasting Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import json
+import requests
+from requests.exceptions import HTTPError
+try:
+    from urlparse import urljoin
+except ImportError:
+    from urllib.parse import urljoin
+
+from ..mdnsbridge import IppmDNSBridge
+
+NMOSOAUTH_DIR = '/var/nmosoauth'
+CREDENTIALS_FILE = 'oauth_credentials.json'
+CREDENTIALS_PATH = os.path.join(NMOSOAUTH_DIR, CREDENTIALS_FILE)
+
+MDNS_SERVICE_TYPE = "nmos-auth"
+
+API_NAMESPACE = 'x-nmos/auth/v1.0/'
+REGISTRATION_ENDPOINT = API_NAMESPACE + 'register_client'
+TOKEN_ENDPOINT = API_NAMESPACE + 'token'
+
+
+def get_credentials_from_file(file):
+    with open(CREDENTIALS_PATH, 'r') as f:
+        credentials = json.load(f)
+    client_id = credentials['client_id']
+    client_secret = credentials['client_secret']
+    return client_id, client_secret
+
+
+class AuthRegistrar(object):
+    def __init__(self, client_name, redirect_uri, client_uri=None,
+                 allowed_scope=None, allowed_grant="authorization_code",
+                 allowed_response="code", auth_method="client_secret_basic",
+                 certificate=None):
+        self.client_name = client_name
+        self.redirect_uri = redirect_uri
+        self.client_uri = client_uri
+        self.allowed_scope = allowed_scope
+        self.allowed_grant = allowed_grant
+        self.allowed_response = allowed_response
+        self.auth_method = auth_method
+        self.certificate = certificate  # Could be deployed centrally for an alternative trust mechanism
+
+        self.bridge = IppmDNSBridge()
+        self.initialised = self.initialise()
+        self.client_id = None
+        self.client_secret = None
+
+    def write_credentials_to_file(self, data, file_path):
+        self.client_id = data.get('client_id')
+        self.client_secret = data.get('client_secret')
+        credentials = {"client_id": self.client_id, "client_secret": self.client_secret}
+        with open(file_path, 'w') as f:
```

Comment:
It would be good to add something here that ensures that the file is only readable by the current user. There will be a way in Python to set file permissions somewhere.